# to qsub snakemake:
#source activate snakemake
#briR
#module load 

#qsub -q short.q -N ChIP_gayther_2015_FT246 -b y -wd /share/ScratchGeneral/jamtor/projects/hgsoc_repeats/histone-ChIP-seq/ -j y -R y -pe smp 13 -V \
#"snakemake -s /share/ScratchGeneral/jamtor/projects/hgsoc_repeats/histone-ChIP-seq/Snakefile \
#--cores 12"

import os
import re

exp_name = 'gayther_2015_FT246'

home_dir = '/share/ScratchGeneral/jamtor/'
project_dir = home_dir + 'projects/hgsoc_repeats/histone-ChIP-seq/'
fastq_dir = 'raw_files/fastq/' + exp_name + '/'
if not os.path.exists(project_dir + 'raw_files/fastq/' + exp_name + '/'):
	os.makedirs(project_dir + 'raw_files/fastq/' + exp_name + '/')

SAMPLES = [ re.sub('\\_pass.fastq.gz', '', x) for x in list(os.walk(fastq_dir))[0][2] ]

print(SAMPLES)

input_sample = 'FT246_input_1_SRR1983929'

# define/create directories:
sra_dir = 'raw_files/sra/' + exp_name + '/'

fastqc_dir = 'results/fastqc/' + exp_name + '/'
if not os.path.exists(project_dir + 'results/fastqc/' + exp_name + '/'):
	os.makedirs(project_dir + 'results/fastqc/' + exp_name + '/')

bwa_dir = 'results/bwa/' + exp_name + '/'
if not os.path.exists(project_dir + 'results/bwa/' + exp_name + '/'):
	os.makedirs(project_dir + 'results/bwa/' + exp_name + '/')

spp_dir = 'results/spp/' + exp_name + '/'
if not os.path.exists(project_dir + 'results/spp/' + exp_name + '/'):
	os.makedirs(project_dir + 'results/spp/' + exp_name + '/')

macs_dir = 'results/macs2/' + exp_name + '/'
if not os.path.exists(project_dir + 'results/macs2/' + exp_name + '/'):
	os.makedirs(project_dir + 'results/macs2/' + exp_name + '/')

# convert all SAMPLE sras to fastqs and move to correct directory:
#for s in SAMPLES:
#	if os.path.isfile(project_dir + 'raw_files/fastq/' + exp_name + 
#		'/' + s + '_pass.fastq.gz'):
#		print(project_dir + 'raw_files/fastq/' + exp_name + 
#		'/' + s + '_pass.fastq.gz already exists')
#	else:
#		print('Creating ' + project_dir + 'raw_files/fastq/' + 
#			exp_name + '/' + s + '.fastq.gz')
#		os.system('fastq-dump --outdir ' + project_dir + 
#			'raw_files/fastq/' + exp_name + '/' + ' --gzip ' + 
#			'--skip-technical --readids --read-filter pass ' + 
#			'--dumpbase --split-3 --clip ' + sra_dir + '/' + s + 
#			'.sra')

# add only narrow peak samples to list for macs_narrow job:
for i in range(0,len(SAMPLES)):
	print(i)
	if 'H3K4' in SAMPLES[i] or 'H3K27ac' in SAMPLES[i]:
		print('is narrow peak H3K4 or H3K27ac sample')
		if 'nar' not in globals():
			nar = [SAMPLES[i]]
		else:
			nar.append(SAMPLES[i])

# add only broad peak samples to list for macs_narrow job:
for i in range(0,len(SAMPLES)):
	print(i)
	if 'H3K27me3' in SAMPLES[i] or 'H3K9me3' in SAMPLES[i]:
		print('is broad peak H3K27me3 or H3K9me3 sample')
		if 'brd' not in globals():
			brd = [SAMPLES[i]]
		else:
			brd.append(SAMPLES[i])

if 'nar' in globals() and 'brd' in globals():
	print('Both narrow and broak peaks samples exist')
	rule all:
		input:
			expand(macs_dir + '{sample}_peaks.narrowPeak', sample=nar),
			expand(macs_dir + '{sample}_peaks.broadPeak', sample=brd),
			expand(bwa_dir + 'all_bams_exist'),
			expand(bwa_dir + '{sample}.sorted.rmdup.bam.bai', sample=SAMPLES),
			#expand(spp_dir + '{sample}.spp.result', sample=SAMPLES),
			expand(bwa_dir + '{sample}.sorted.bam.bai', sample=SAMPLES),
			#expand('genomes/repeats/bwa/hg38_ercc_all_repbase.bwa.amb', sample=SAMPLES),
			#expand(fastqc_dir + '{sample}_pass_fastqc.html', sample=SAMPLES)

elif 'nar' in globals() and 'brd' not in globals():
	print('Only narrow peak samples exist')
	rule all:
		input:
			expand(macs_dir + '{sample}_peaks.narrowPeak', sample=nar),
			#expand(bwa_dir + 'all_bams_exist'),
			expand(bwa_dir + '{sample}.sorted.rmdup.bam.bai', sample=SAMPLES),
			#expand(spp_dir + '{sample}.spp.result', sample=SAMPLES),
			expand(bwa_dir + '{sample}.sorted.bam.bai', sample=SAMPLES),
			#expand('genomes/repeats/bwa/hg38_ercc_all_repbase.bwa.amb', sample=SAMPLES),
			#expand(fastqc_dir + '{sample}_pass_fastqc.html', sample=SAMPLES)

elif 'brd' in globals() and 'nar' not in globals():
	print('Only broak peak samples exist')
	rule all:
		input:
			expand(macs_dir + '{sample}_peaks.broadPeak', sample=brd),
			#expand(bwa_dir + 'all_bams_exist'),
			expand(bwa_dir + '{sample}.sorted.rmdup.bam.bai', sample=SAMPLES),
			#expand(spp_dir + '{sample}.spp.result', sample=SAMPLES),
			expand(bwa_dir + '{sample}.sorted.bam.bai', sample=SAMPLES),
			#expand('genomes/repeats/bwa/hg38_ercc_all_repbase.bwa.amb', sample=SAMPLES),
			#expand(fastqc_dir + '{sample}_pass_fastqc.html', sample=SAMPLES)

#rule fastqc:
#	input:
#		fastq_dir + '{sample}_pass.fastq.gz'
#	output:
#		fastqc_dir + '{sample}_pass_fastqc.html',
#		fastqc_dir + '{sample}_pass_fastqc.zip'
#	shell:
#		'fastqc -t 2 -o ' + fastqc_dir + ' {input}'

#rule bwa_index:
#	input:
#		'genomes/repeats/hg38_ercc_all_repbase.fa'
#	output:
#		'genomes/repeats/bwa/hg38_ercc_all_repbase.bwa.amb'
#	shell:
#		'bwa index -p genomes/repeats/bwa/hg38_ercc_all_repbase.bwa {input}'

rule bwa_align:
	input:
		fasq = fastq_dir + '{sample}_pass.fastq.gz',
#		fqc = fastqc_dir + '{sample}_pass_fastqc.html'
	output:
		temp(bwa_dir + '{sample}.sai')
	threads: 8
	shell:
		'bwa aln -f {output} genomes/repeats/bwa/hg38_ercc_all_repbase.bwa {input.fasq}'

rule bwa_samse:
	input:
		sai = bwa_dir + '{sample}.sai',
		fq = fastq_dir + '{sample}_pass.fastq.gz'
	output:
		temp(bwa_dir + '{sample}.sam')
	shell:
		'bwa samse -f {output} genomes/repeats/bwa/hg38_ercc_all_repbase.bwa \
			{input.sai} {input.fq}'

rule samtools_filter:
	input:
		bwa_dir + '{sample}.sam'
	output:
		bwa_dir + '{sample}.sorted.bam'
	shell:
		'samtools view -u -F260 {input}  | samtools sort -o {output} \
		-O bam -T ' + bwa_dir + '{wildcards.sample}.tmp -@ 2 -'


rule samtools_index:
	input:
		bwa_dir + '{sample}.sorted.bam'
	output:
		temp(bwa_dir + '{sample}.sorted.bam.bai')
	shell:
		'samtools index {input}'

rule remove_dupes:
	input:
		bam = bwa_dir + '{sample}.sorted.bam',
		bai = bwa_dir + '{sample}.sorted.bam.bai'
	output:
		rmdup = bwa_dir + '{sample}.sorted.rmdup.bam',
		met = temp(bwa_dir + '{sample}.sorted.rmdupmet.bam')
	threads: 2
	shell:
		'java -jar /home/jamtor/local/lib/picard/build/libs/picard.jar \
		MarkDuplicates I={input.bam} O={output.rmdup} M={output.met} \
		REMOVE_DUPLICATES=true VALIDATION_STRINGENCY=LENIENT'

rule samtools_index2:
	input:
		bwa_dir + '{sample}.sorted.rmdup.bam'
	output:
		bwa_dir + '{sample}.sorted.rmdup.bam.bai'
	shell:
		'samtools index {input}'


#rule spp:
#	input:
#		bwa_dir + '{sample}.sorted.rmdup.bam'
#	output:
#		spp_dir + '{sample}.spp.result'
#	shell:
#		'Rscript /home/jamtor/local/lib/phantompeakqualtools/run_spp.R \
#		-c={input} -i={input} -savp -p=2 -odir=' + spp_dir +
#		' -out={output}'

# make following rules wait until all outputs are complete:
#checkpt = [ bwa_dir + x + 
#	'.sorted.rmdup.bam' for x in [s for s in SAMPLES if 'input' in s] ]
#
#input_checkpt = ''
#for i in range(0, len(checkpt)):
#	input_checkpt = input_checkpt + 'file' + str(i) + "=" + checkpt[i] + ', '
#input_checkpt = re.sub('bam, $', 'bam', input_checkpt)
#
#rule check_files:
#	input:
#		input_checkpt
#	output:
#		bwa_dir + 'all_bams_exist'
#	shell:
#		'touch {output}'


rule macs_narrow:
	input:
		#chek = bwa_dir + 'all_bams_exist',
		rmdup_bam = bwa_dir + '{sample}.sorted.rmdup.bam',
		input_bam = bwa_dir + input_sample + '.sorted.rmdup.bam'
	output:
		macs_dir + '{sample}_peaks.narrowPeak'
	shell:
		'source activate p2.7env; ' +
		'/home/jamtor/local/lib/MACS/bin/macs2 callpeak -t ' +
		'{input.rmdup_bam} -c {input.input_bam} --name ' +
		'"{wildcards.sample}" --outdir ' + macs_dir

rule macs_broad:
	input:
		chek = bwa_dir + 'all_bams_exist',
		rmdup_bam = bwa_dir + '{sample}.sorted.rmdup.bam'
	output:
		macs_dir + '{sample}_peaks.broadPeak'
	shell:
		'source activate p2.7env; ' +
		'/home/jamtor/local/lib/MACS/bin/macs2 callpeak -t ' +
		'{input.rmdup_bam} -c ' +
		bwa_dir + input_sample + '.sorted.rmdup.bam' +
		' --broad' +
		' --name "{wildcards.sample}" ' +
		'--outdir ' + macs_dir
